src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,249 @@
+                          const Principal& principal,
+                          const Predicate& predicate) {
+    return PushOntoPredicate(modifier, {principal.name()}, predicate);
+  }
+
+  Predicate AttributeToDLIR(const Attribute& attribute);
+
+  // Translating an attribute results in two objects:
+  // - A predicate, this is the main result of the translation
+  // - An additional rule that allows attributes to be passed with canActAs
+  // This function is for generating the additional rule.
+  // This takes both a speaker and the actual attribute as argument because
+  // the speaker is needed to generate the additional rule. Note that all base
+  // facts have a speaker either explicitly or implicitly:
+  // - If it appears on the LHS of an assertion, it explicitly has a speaker
+  // - If it appears on the RHS of an assertion, it behaves semantically
+  // like it has the same speaker as the head of the assertion.
+  DLIRAssertion SpokenAttributeToDLIR(const Principal& speaker,
+                                      const Attribute& attribute);
+
+  Predicate CanActAsToDLIR(const CanActAs& can_act_as);
+
+  // In the same way that attributes are passed around with "CanActAs", so
+  // are other "CanActAs" facts. To implement this, the translation of
+  // CanActAs also results in both a predicate and an extra rule that passes
+  // these facts around. This function is for generating the extra rule and
+  // it works similarly to "SpokenAttributeToDLIR".
+  DLIRAssertion SpokenCanActAsToDLIR(const Principal& speaker,
+                                     const CanActAs& can_act_as);
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      const Principal& speaker, const Predicate& predicate);
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      const Principal& speaker, const Attribute& attribute);
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      const Principal& speaker, const CanActAs& canActAs);
+
+  // The second return value represents 0 or 1 newly generated rules, so an
+  // option might seem more intuitive. However, the interface that consumes
+  // this needs to construct a vector anyway, so a vector is used in the
+  // return type.
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIR(
+      const Principal& speaker, const BaseFact& base_fact);
+
+  // This can result in 0 or more new rules because the translation of
+  // nested canSayFacts might result in more than 1 rule.
+  std::pair<Predicate, std::vector<DLIRAssertion>> FactToDLIR(
+      const Principal& speaker, const Fact& fact);
+
+  std::vector<DLIRAssertion> SingleSaysAssertionToDLIR(
+      const Principal& speaker, const Assertion& assertion);
+
+  std::vector<DLIRAssertion> SaysAssertionToDLIR(
+      const SaysAssertion& says_assertion);
+
+  std::vector<DLIRAssertion> SaysAssertionsToDLIR(
+      const std::vector<SaysAssertion>& says_assertions);
```

Comment:
A lot of these do actually need to be member functions because they either directly call `FreshVar()` which is stateful or they call some other member that calls `FreshVar()`. I did remove the ones that are not stateful (which includes the `PushPredicate` helpers and the two base fact variants.)